src/core/sync/event.d

```diff
@@ -0,0 +1,318 @@
+/**
+ * The event module provides a primitive for lightweight signalling of other threads
+ * (aka Windows Events)
+ *
+ * Copyright: Copyright (c) 2019 D Language Foundation
+ * License: Distributed under the
+ *    $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0).
+ *    (See accompanying file LICENSE)
+ * Authors: Rainer Schuetze
+ * Source:    $(DRUNTIMESRC core/sync/event.d)
+ */
+module core.sync.event;
+
+version (Windows)
+{
+    import core.sys.windows.basetsd /+: HANDLE +/;
+    import core.sys.windows.winerror /+: WAIT_TIMEOUT +/;
+    import core.sys.windows.winbase /+: CreateEvent, CloseHandle, SetEvent, ResetEvent,
+        WaitForSingleObject, INFINITE, WAIT_OBJECT_0+/;
+}
+else version (Posix)
+{
+    import core.sys.posix.pthread;
+    import core.sys.posix.sys.types;
+}
+else
+{
+    static assert(false, "Platform not supported");
+}
+
+import core.time;
+import core.internal.abort : abort;
+
+/**
+ * This class represents an event. Clients of an event are suspended while waiting
+ * for the event to be "signaled".
+ *
+ * Implemented using `pthread_mutex` and `pthread_condition` on Posix and
+ * `CreateEvent` and `SetEvent` on Windows.
+ */
+struct Event
+{
+    // Posix version inspired by http://www.it.uu.se/katalog/larme597/win32eoposix
+nothrow @nogc:
+    /**
+     * Creates an event object.
+     *
+     * Params:
+     *  manualReset  = the state of the event is not reset automatically after resuming waiting clients
+     *  initialState = initial state of the signal
+     */
+    this(bool manualReset, bool initialState)
+    {
+        initialize(manualReset, initialState);
+    }
+
+    /**
+     * Initializes an event object.
+     *
+     * Params:
+     *  manualReset  = the state of the event is not reset automatically after resuming waiting clients
+     *  initialState = initial state of the signal
+     */
+    void initialize(bool manualReset, bool initialState)
```

Comment:
Why is this a separate method?

Reply:
You cannot pass arguments to the constructor if it is part of another struct, e.g. https://github.com/dlang/druntime/pull/2514/files#diff-6615611ba51118215a271e3500c61122R2605. You have to use an explicit `initialize`-call then. `__ctor` could be an alternative, but I think it should not be part of the API.

Reply:
Is there a downside to do `evStart = Event(...)` in `Gcx.initialize`?

Reply:
I've explicitly added `@disable this(this)` as this could duplicate the handle reference.

Reply:
Does not seem to prevent assignment: https://run.dlang.io/is/iefkBG.

Reply:
Hmm, seems like opAssign should be disabled, too.

Reply:
If that's required then I'm fine with using `initialize` instead of a constructor.